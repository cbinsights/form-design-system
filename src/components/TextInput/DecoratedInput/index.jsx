import React from 'react';
import PropTypes from 'prop-types';

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const DecoratedInput = ({ before, after, children }) => (
  <div className="fdsDecoratedInput" data-before={before} data-after={after}>
    {children}
  </div>
);

DecoratedInput.propTypes = {
  /** String to place to the left of the input */
  before: PropTypes.string,

  /** String to place to the right of the input */
  after: PropTypes.string,

  /** React children */
  children: PropTypes.node,
};

export default DecoratedInput;
