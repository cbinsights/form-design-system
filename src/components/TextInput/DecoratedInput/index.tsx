import React from 'react';

export interface DecoratedInputProps {
  /** String to place to the left of the input */
  before?: string;

  /** String to place to the right of the input */
  after?: string;

  /** React children */
  children?: Array<React.ReactNode> | React.ReactNode;
}

const DecoratedInput = ({
  before,
  after,
  children,
}: DecoratedInputProps): JSX.Element => (
  <div className="fdsDecoratedInput" data-before={before} data-after={after}>
    {children}
  </div>
);

export default DecoratedInput;
