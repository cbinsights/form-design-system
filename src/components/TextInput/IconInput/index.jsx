import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';

import FDS from 'lib/dictionary/js/styleConstants';

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const IconInput = ({ IconLeft, IconRight, children }) => {
  const classNames = cx('fdsIconInput', {
    'fdsIconInput--hasStartIcon': Boolean(IconLeft),
    'fdsIconInput--hasEndIcon': Boolean(IconRight),
  });

  return (
    <div className={classNames}>
      {IconLeft && (
        <span className="fdsIconInput-icon fdsIconInput-icon--start alignChild--center--center">
          <IconLeft size="xs" color={FDS.FONT_COLOR_SECONDARY} />
        </span>
      )}
      {children}
      {IconRight && (
        <span className="fdsIconInput-icon fdsIconInput-icon--end alignChild--center--center">
          <IconRight size="xs" color={FDS.FONT_COLOR_SECONDARY} />
        </span>
      )}
    </div>
  );
};

IconInput.propTypes = {
  /** An `input` element child */
  children: PropTypes.node,

  /**
   * FDS Icon _reference_
   * `IconLeft={StarFilledIcon}`
   */
  IconLeft: PropTypes.func,

  /**
   * FDS Icon _reference_
   * `IconRight={StarFilledIcon}`
   */
  IconRight: PropTypes.func,
};

export default IconInput;
