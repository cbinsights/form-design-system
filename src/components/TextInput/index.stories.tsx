import React from 'react';
import CaretDownIcon from 'icons/react/CaretDownIcon';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import { useForm } from 'react-hook-form';
import InputGroup from 'components/InputGroup';
import State from 'util/storybook-docs/State';
import TextInput, { TextInputProps } from '.';
import { noop } from 'util/index';
import { StoryObj } from '@storybook/react';

export const Primary: StoryObj<TextInputProps> = {
  args: {
    label: 'First Name',
  },
};

export const ExampleLayout: StoryObj<TextInputProps> = {
  render: ({ onChange }: TextInputProps): JSX.Element => (
    <>
      <Flex>
        <FlexItem>
          <TextInput label="First Name" showRequired onChange={onChange} />
        </FlexItem>
        <FlexItem>
          <TextInput label="Last Name" showRequired onChange={onChange} />
        </FlexItem>
      </Flex>
      <br />
      <Flex>
        <FlexItem>
          <TextInput
            type="email"
            label="Email"
            errorText="Enter a valid email address"
            placeholder="e.g. anand@cbinsights.com"
            onChange={onChange}
          />
        </FlexItem>
        <FlexItem>
          <TextInput type="password" label="Password" onChange={onChange} />
        </FlexItem>
        <FlexItem>
          <TextInput type="number" label="Zipcode" after="90210" onChange={onChange} />
        </FlexItem>
      </Flex>
    </>
  ),
};

export const ControlledInput: StoryObj<TextInputProps> = {
  render: ({ onChange = noop, ...args }: TextInputProps): JSX.Element => (
    <State
      initialValue="Controlled"
      render={(value, setValue) => (
        <TextInput
          {...args}
          label="Controlled"
          showLabel={false}
          value={value}
          onChange={(e) => {
            onChange(e);
            setValue(e.target.value);
          }}
        />
      )}
    />
  ),
};

export const Icons: StoryObj<TextInputProps> = {
  args: {
    label: 'Icons',
    showLabel: false,
    IconLeft: CaretDownIcon,
    IconRight: CaretDownIcon,
  },
};

export const SideLabel: StoryObj<TextInputProps> = {
  args: {
    label: 'Side Label',
    showLabel: false,
    before: 'before',
    after: 'after',
  },
};

export const FormManagementLibrary: StoryObj<TextInputProps> = {
  render: ({ onChange, ...args }: TextInputProps): JSX.Element => {
    const { register, handleSubmit, formState } = useForm();
    const registeredUsername = register('Username', {
      validate: (value) => value !== 'admin' || 'Nice try!',
    });
    const registeredEmail = register('Email', {
      required: 'Required',
      pattern: {
        value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i,
        message: 'invalid email address',
      },
    });
    return (
      <div>
        <form onSubmit={handleSubmit(() => {})}>
          <div className="margin--bottom--s">
            <TextInput
              {...args}
              label="Username"
              name="username"
              type="text"
              errorText={formState.errors.username && formState.errors.username.message}
              ref={registeredUsername.ref}
              onChange={onChange}
            />
          </div>
          <TextInput
            {...args}
            label="Email"
            name="email"
            type="text"
            errorText={formState.errors.email && formState.errors.email.message}
            ref={registeredEmail.ref}
            onChange={onChange}
          />
          <br />
          <button type="submit">Submit</button>
        </form>
      </div>
    );
  },
  parameters: {
    docs: {
      description: {
        story:
          'You may use this input in any form management library you would like. We recommend `react-hook-form`, which is used in the example below:',
      },
    },
  },
};

export const Misc: StoryObj<TextInputProps> = {
  render: ({ onChange }: TextInputProps): JSX.Element => (
    <>
      <InputGroup>
        <TextInput
          before="$"
          type="number"
          defaultValue={666}
          label="Misc"
          showLabel={false}
          onChange={onChange}
        />
        <select name="unit">
          <option value="1">Million</option>
          <option value="2">Thousand</option>
        </select>
      </InputGroup>
      <br />
      <InputGroup>
        <TextInput
          IconLeft={CaretDownIcon}
          type="number"
          defaultValue={666}
          label="Misc"
          showLabel={false}
          onChange={onChange}
        />
        <select name="unit">
          <option value="1">Million</option>
          <option value="2">Thousand</option>
        </select>
      </InputGroup>
      <br />
      <InputGroup>
        <TextInput
          IconRight={CaretDownIcon}
          type="number"
          defaultValue={666}
          label="Misc"
          showLabel={false}
          onChange={onChange}
        />
        <select name="unit">
          <option value="1">Million</option>
          <option value="2">Thousand</option>
        </select>
      </InputGroup>
      <br />
      <TextInput
        label="Label on left with width"
        labelPosition="left"
        labelWidth="100px"
        onChange={onChange}
      />
    </>
  ),
};

export default {
  component: TextInput,
  title: 'Components/TextInput',
  argTypes: {
    onChange: {
      action: 'onChange',
    },
  },
  parameters: {
    componentSubtitle: 'Standard recommended input component',
    docs: {
      fds: {
        attributes: ['rest'],
      },
      description: {
        component:
          'Can be used for simple no frills inputs, as well as complex inputs complete with accessibile labels, Icons, side labels, errorText, throttled events, and more.',
      },
    },
  },
};
