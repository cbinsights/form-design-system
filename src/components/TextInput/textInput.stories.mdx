import { Meta, Story, Preview, Props, Source } from "@storybook/addon-docs/blocks";
import { withKnobs, text, boolean, optionsKnob as options } from '@storybook/addon-knobs';
import CaretDownIcon from 'lib/icons/react/CaretDownIcon';
import { Info, State, arrayToOptions } from 'components/util/storybook';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import { action } from "@storybook/addon-actions";
import { useForm } from "react-hook-form";

import { ImportPath } from 'components/util/storybook';
import TextInput from '.';
import InputGroup from 'components/InputGroup';

<Meta title="Components/TextInput" decorators={[withKnobs]} component={TextInput} />

# TextInput

<ImportPath component={TextInput} section="forms" />

Standard recommended input component. Can be used for simple no frills inputs, as well as complex inputs
complete with accessibile labels, Icons, side labels, errorText, throttled events, and more.

<Story name="Knobs" parameters={{ docs: { disable: true } }}>
  <TextInput
    label={text("label", "Label")}
    errorText={text("errorText", undefined)}
    showRequired={boolean("required", undefined)}
    type={text("type", undefined)}
    onChange={action("changed")}
    IconRight={boolean("Show IconRight", undefined) ? CaretDownIcon : undefined}
    IconLeft={boolean("Show IconLeft", undefined) ? CaretDownIcon : undefined}
    before={text("before", undefined)}
    after={text("after", undefined)}
    labelPosition={options(
      "labelPosition",
      arrayToOptions(['top', 'left']),
      undefined,
      { display: "inline-radio" }
    )}
  />
</Story>

<Info type="info">
  <b>Side labels are not a substitute for label elements</b>
</Info>

<Preview isExpanded>
  <>
    <TextInput label="FirstName" showLabel={false} />
    <br />
    <TextInput
      label="First Name"
      showRequired
      errorText="Custom error text"
      placeholder="Placeholder text"
      IconRight={CaretDownIcon}
      before="$"
      after=".com"
    />
  </>
</Preview>

<Info type="info">
  Additional props will be spread on the `input` element.
</Info>

<Props of={TextInput} />

## Example Layout

<Preview>
  <Story name="Example Layout">
    <>
      <Flex>
        <FlexItem>
          <TextInput label="First Name" showRequired />
        </FlexItem>
        <FlexItem>
          <TextInput label="Last Name" showRequired />
        </FlexItem >
      </Flex>
      <br />
      <Flex>
        <FlexItem>
          <TextInput
            type="email"
            label="Email"
            errorText="Enter a valid email address"
            placeholder="e.g. anand@cbinsights.com"
          />
        </FlexItem>
        <FlexItem>
          <TextInput type="password" label="Password" />
        </FlexItem>
        <FlexItem>
          <TextInput type="number" label="Zipcode" after="90210" />
        </FlexItem>
      </Flex>
    </>
  </Story>
</Preview>

### Throttle

The prop `onThrottleChange` fires a callback that throttles less than 5 characters,
and debounces 5 or more characters.
https://www.peterbe.com/plog/how-to-throttle-and-debounce-an-autocomplete-input-in-react

View this story in Canvas, in the action tab, to see how the callback is throttled.

<Preview>
  <Story name="Throttle">
    <TextInput
      onThrottledChange={action("changed")}
      label="Throttled"
      showLabel={false}
    />
  </Story>
</Preview>


### Controlled Input

<Preview>
  <Story name="Controlled Input">
    <State
      initialValue="Controlled"
      render={(value, setValue) => (
        <TextInput
          label="Controlled"
          showLabel={false}
          value={value}
          onChange={(e) => setValue(e.target.value)}
        />
      )}
    />
  </Story>
</Preview>

### Icons

<Preview>
  <Story name="Icons">
    <TextInput
      label="Icons"
      showLabel={false}
      IconLeft={CaretDownIcon}
      IconRight={CaretDownIcon}
    />
  </Story>
</Preview>

### Side label

<Preview>
  <Story name="Before & After">
    <TextInput
      label="Side Label"
      showLabel={false}
      before="before"
      after="after"
    />
  </Story>
</Preview>

### Side Labels + Icon

<Preview>
  <Story name="Side + Icon">
    <TextInput
      label="Label"
      placeholder="placeholder"
      before="before"
      after="after"
      IconLeft={CaretDownIcon}
      IconRight={CaretDownIcon}
    />
  </Story>
</Preview>

### Form management library
You may use this input in any form management library you'd like. We recommend `react-hook-form`, which is used in the example below:

<Preview>
  <Story name="Form management library">
    {() => {
      const { register, handleSubmit, setValue, errors } = useForm();
      return <div>
        <form onSubmit={handleSubmit(action('submitted'))}>
          <div className="margin--bottom--half">
          <TextInput
            label="Username"
            name="username"
            type="text"
            errorText={errors.username && errors.username.message}
            ref={register({
              validate: value => value !== "admin" || "Nice try!"
            })}
          />
          </div>
          <TextInput
            label="Email"
            name="email"
            type="text"
            errorText={errors.email && errors.email.message}
            ref={register({
              required: 'Required',
              pattern: {
                value: /^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i,
                message: "invalid email address"
              }
            })}
          />
          <br />
          <button type="submit">Submit</button>
        </form>
      </div>;
    }}
  </Story>
</Preview>

### Misc

Can be used with InputGroup

<Preview>
  <Story name="Misc">
    <>
      <InputGroup>
        <TextInput before="$" type="number" value="666" label="Misc" showLabel={false} />
        <select name="unit">
          <option value="1">Million</option>
          <option value="2">Thousand</option>
        </select>
      </InputGroup>
      <br />
      <InputGroup>
        <TextInput IconLeft={CaretDownIcon} type="number" value="666" label="Misc" showLabel={false} />
        <select name="unit">
          <option value="1">Million</option>
          <option value="2">Thousand</option>
        </select>
      </InputGroup>
      <br />
      <InputGroup>
        <TextInput IconRight={CaretDownIcon} type="number" value="666" label="Misc" showLabel={false}/>
        <select name="unit">
          <option value="1">Million</option>
          <option value="2">Thousand</option>
        </select>
      </InputGroup>
    </>
  </Story>
</Preview>
