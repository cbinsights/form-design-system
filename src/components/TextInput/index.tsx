import React, { RefObject } from 'react';
import cc from 'classcat';
import IconInput from './IconInput';
import DecoratedInput from './DecoratedInput';
import { IconProps } from 'components/Icon';

export interface TextInputProps
  extends React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  > {
  /** Control aria-label */
  'aria-label'?: string;
  /** Control whether input has error styling */
  hasError?: boolean;
  /** Control whether label receives a bold treatment */
  isLabelBold?: boolean;
  /** Displays as red error text below input */
  errorText?: string;

  /** Label used for `label` element */
  label: string;

  /**
   * FDS Icon _reference_
   * `IconLeft={StarFilledIcon}`
   */
  IconLeft?: React.ComponentType<IconProps>;

  /**
   * FDS Icon _reference_
   * `IconRight={StarFilledIcon}`
   */
  IconRight?: React.ComponentType<IconProps>;

  /** String to place to the left of the input. Not a substitute for a label! */
  before?: string;

  /** String to place to the right of the input. Not a substitute for a label! */
  after?: string;

  /** Controls where the label renders */
  labelPosition?: 'top' | 'left';

  /** Specify width of label (useful for when label is positioned to left */
  labelWidth?: string;

  showRequired?: boolean;
  showLabel?: boolean;
  ref?: RefObject<HTMLInputElement>;
}

const TextInput = React.forwardRef<HTMLInputElement, TextInputProps>(
  (
    {
      label,
      labelWidth,
      isLabelBold,
      labelPosition = 'top',
      errorText,
      hasError,
      showRequired,
      showLabel = true,
      onChange = () => {},
      type = 'text',
      IconLeft,
      IconRight,
      after,
      before,
      'aria-label': ariaLabel,
      ...rest
    }: TextInputProps,
    ref
  ) => {
    const Element = showLabel ? 'label' : 'div';

    return (
      <>
        <Element
          className={cc([
            'fdsTextInput-root',
            `fdsTextInput-root--${labelPosition}Label`,
          ])}
        >
          {showLabel && label && (
            <div
              className={cc([
                {
                  'fontWeight--bold': isLabelBold,
                  'margin--bottom--xs': labelPosition === 'top',
                },
                'fdsTextInput-label',
              ])}
              style={labelWidth ? { width: labelWidth } : undefined}
            >
              {label}
              {showRequired && <span className="color--red">&nbsp;*</span>}
            </div>
          )}
          <DecoratedInput after={after} before={before}>
            <IconInput IconLeft={IconLeft} IconRight={IconRight}>
              <input
                {...rest}
                ref={ref}
                aria-label={ariaLabel || (showLabel ? label : undefined)}
                aria-invalid={errorText || hasError ? true : undefined}
                onChange={onChange}
                type={type}
                className={cc([
                  {
                    error: errorText || hasError,
                  },
                  'fdsTextInput',
                ])}
              />
            </IconInput>
          </DecoratedInput>
        </Element>
        {errorText && <div className="color--red type--data">{errorText}</div>}
      </>
    );
  }
);

TextInput.displayName = 'TextInput';
export default TextInput;
