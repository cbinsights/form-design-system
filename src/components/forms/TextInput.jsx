import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { throttle, debounce } from 'throttle-debounce';
import IconInput from 'components/forms/IconInput';
import DecoratedInput from 'components/forms/DecoratedInput';
import cc from 'classcat';

export const throttleValue = 500;

export const throttleInput = (input, throttleFn, debounceFn) => {
  if (input.length < 5 || input.endsWith(' ')) {
    throttleFn(input);
  } else {
    debounceFn(input);
  }
};

const TextInput = React.forwardRef(
  (
    {
      label,
      labelPosition = 'top',
      errorText,
      onThrottledChange,
      showRequired,
      showLabel = true,
      onChange = () => {},
      type = 'text',
      IconLeft,
      IconRight,
      after,
      before,
      ...props
    },
    ref
  ) => {
    // Refer to the following for an explanation on
    // the throttling / debouncing approach:
    // https://www.peterbe.com/plog/how-to-throttle-and-debounce-an-autocomplete-input-in-react
    let onThrottledChangeThrottled;
    let onThrottledChangeDebounced;

    const assignThrottleFunctions = () => {
      if (onThrottledChange) {
        onThrottledChangeThrottled = throttle(throttleValue, onThrottledChange);
        onThrottledChangeDebounced = debounce(throttleValue, onThrottledChange);
      }
    };

    const inputOnChange = (e) => {
      e.persist();
      onChange(e);
      if (onThrottledChange) {
        throttleInput(
          e.target.value,
          onThrottledChangeThrottled,
          onThrottledChangeDebounced
        );
      }
    };

    useEffect(assignThrottleFunctions, [onThrottledChange]);

    const Element = showLabel ? 'label' : 'div';

    return (
      <>
        <Element
          className={cc([
            'fdsTextInput-root',
            `fdsTextInput-root--${labelPosition}Label`,
          ])}
        >
          {showLabel && label && (
            <div className="fdsTextInput-label">
              {label}
              {showRequired && <span className="color--red">&nbsp;*</span>}
            </div>
          )}
          <DecoratedInput after={after} before={before}>
            <IconInput IconLeft={IconLeft} IconRight={IconRight}>
              <input
                {...props}
                ref={ref}
                aria-label={showLabel ? label : undefined}
                onChange={inputOnChange}
                type={type}
                className={cc([
                  {
                    error: errorText,
                  },
                  'fdsTextInput',
                ])}
              />
            </IconInput>
          </DecoratedInput>
        </Element>
        {errorText && <div className="color--red type--data">{errorText}</div>}
      </>
    );
  }
);

TextInput.displayName = 'TextInput';

TextInput.propTypes = {
  errorText: PropTypes.string,

  /** Label used for `label` element */
  label: PropTypes.string.isRequired,

  /**
   * FDS Icon _reference_
   * `IconLeft={StarFilledIcon}`
   */
  IconLeft: PropTypes.func,

  /**
   * FDS Icon _reference_
   * `IconRight={StarFilledIcon}`
   */
  IconRight: PropTypes.func,

  /** Input element `type` attribute */
  type: PropTypes.string,

  /** Controls showing asterisk on label, signifying field is required */
  required: PropTypes.bool,

  /** Standard React onChange event */
  onChange: PropTypes.func,

  /** Custom onChange event that throttles / debounces. */
  onThrottledChange: PropTypes.func,

  /** String to place to the left of the input */
  before: PropTypes.string,

  /** String to place to the right of the input */
  after: PropTypes.string,

  labelPosition: PropTypes.oneOf(['top', 'left']),
  showRequired: PropTypes.bool,
  showLabel: PropTypes.bool,
};

export default TextInput;
