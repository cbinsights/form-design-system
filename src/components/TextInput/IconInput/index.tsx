import React from 'react';
import cc from 'classcat';

import FDS from 'dictionary/js/styleConstants';
import { IconProps } from 'components/Icon';

export interface IconInputProps {
  /** An `input` element child */
  children?: React.ReactNode;

  /**
   * FDS Icon _reference_
   * `IconLeft={StarFilledIcon}`
   */
  IconLeft?: React.ComponentType<IconProps>;

  /**
   * FDS Icon _reference_
   * `IconRight={StarFilledIcon}`
   */
  IconRight?: React.ComponentType<IconProps>;
}

/**
 * @param {Object} props react props
 * @returns {ReactElement}
 */
const IconInput = ({ IconLeft, IconRight, children }: IconInputProps): JSX.Element => {
  const classNames = cc([
    {
      'fdsIconInput--hasStartIcon': Boolean(IconLeft),
      'fdsIconInput--hasEndIcon': Boolean(IconRight),
    },
    'fdsIconInput',
  ]);

  return (
    <div className={classNames}>
      {IconLeft && (
        <span className="fdsIconInput-icon fdsIconInput-icon--start alignChild--center--center">
          <IconLeft size="xs" color={FDS.FONT_COLOR_SECONDARY} />
        </span>
      )}
      {children}
      {IconRight && (
        <span className="fdsIconInput-icon fdsIconInput-icon--end alignChild--center--center">
          <IconRight size="xs" color={FDS.FONT_COLOR_SECONDARY} />
        </span>
      )}
    </div>
  );
};

export default IconInput;
