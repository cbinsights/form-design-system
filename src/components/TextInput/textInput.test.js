import React from 'react';
import { shallow, mount } from 'enzyme';

import TextInput, { throttleInput } from '.';

describe('TextInput component', () => {

  it('fires throttleFunctions', () => {
    const throttleFn = jest.fn()
    const wrapper = mount(<TextInput onThrottledChange={throttleFn} label="label" />);
    const input = wrapper.find('input')
    input.simulate('change', { target: { value: 'aaa' } });
    expect(throttleFn).toHaveBeenCalled();
  });

  it('matches snapshot (default)', () => {
    const wrapper = shallow(<TextInput label="label" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('matches snapshot (all props)', () => {
    const wrapper = shallow(
      <TextInput 
        label="label"
        errorText="errorText"
        IconLeft={() => {}}
        IconRight={() => {}}
        onChange={() => {}}
        onThrottledChange={() => {}}
        before="before"
        after="after"
        type="number"
        labelPosition="left"
        required
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

});

describe('throttleInput', () => {
  let throttleFn;
  let debounceFn

  beforeEach(() => {
    throttleFn = jest.fn();
    debounceFn = jest.fn();
  })

  it('calls throttleFn when less than 5 characters', () => {
    throttleInput('aaa', throttleFn, debounceFn)
    expect(throttleFn).toHaveBeenCalled();
    expect(debounceFn).not.toHaveBeenCalled();
  });

  it('calls throttleFn when a space is the last character', () => {
    throttleInput('aaaaaaaaaaa ', throttleFn, debounceFn)
    expect(throttleFn).toHaveBeenCalled();
    expect(debounceFn).not.toHaveBeenCalled();
  });

  it('calls debounceFn with 5 characters or more', () => {
    throttleInput('aaaaa', throttleFn, debounceFn)
    expect(throttleFn).not.toHaveBeenCalled();
    expect(debounceFn).toHaveBeenCalled();
  });

});
