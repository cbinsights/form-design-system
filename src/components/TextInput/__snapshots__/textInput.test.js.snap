// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput component matches snapshot (all props) 1`] = `
<Fragment>
  <label
    className="fdsTextInput-root fdsTextInput-root--leftLabel"
  >
    <div
      className="fdsTextInput-label"
    >
      label
    </div>
    <DecoratedInput
      after="after"
      before="before"
    >
      <IconInput
        IconLeft={[Function]}
        IconRight={[Function]}
      >
        <input
          aria-label="label"
          className="fdsTextInput error"
          onChange={[Function]}
          required={true}
          type="number"
        />
      </IconInput>
    </DecoratedInput>
  </label>
  <div
    className="color--red type--data"
  >
    errorText
  </div>
</Fragment>
`;

exports[`TextInput component matches snapshot (default) 1`] = `
<Fragment>
  <label
    className="fdsTextInput-root fdsTextInput-root--topLabel"
  >
    <div
      className="fdsTextInput-label"
    >
      label
    </div>
    <DecoratedInput>
      <IconInput>
        <input
          aria-label="label"
          className="fdsTextInput"
          onChange={[Function]}
          type="text"
        />
      </IconInput>
    </DecoratedInput>
  </label>
</Fragment>
`;
